.github/workflows/ci.yaml

```diff
@@ -104,6 +104,21 @@ jobs:
       - run: ./frb_internal generate-website
 
   # ----------------------------------- codegen -----------------------------------
+  msrv:
```

Comment:
nit: just add a few lines below https://github.com/fzyzcjy/flutter_rust_bridge/blob/13c099fe196622f11bf0216affe00d086b867b7b/.github/workflows/ci.yaml#L331-L342

For example

```
          - image: ubuntu-latest
            version: 1.70 # or whatever version you like
```

Reply:
I can do that - but I don't understand completely why ...
cargo MSRV is installing the version specified in frb_codegen/cargo.toml itself, and runs cargo check afterwards. 
If we would specify a specific version in the image, it would get overwritten anyways - and that is good, as we don't need to maintain the rust version on two places (cargo.toml and ci).

Or did you mean something else?

Reply:
Ok I realize we are talking two different things 🤣

* https://github.com/foresterre/cargo-msrv: Find the minimum supported Rust version (MSRV) for your project. Only need to run once locally to find it. 
* MSRV = "minimal supported rust version", e.g. https://internals.rust-lang.org/t/rust-msrv-policy-and-linux-distros/17074. If I understand correctly (correct me if I am wrong!), for example, MSRV=1.60 means "anyone with rust version >=1.60 can happily use my crate as a dependency".

Thus, Suppose we want MSRV=1.70 (surely we can aim lower). Then it seems that we can add a line as suggested above. Then, if the CI fails, it means anyone using rust 1.70 will fail to use our lib, and we should fix errors until ci is happy.

Reply:
Hmm ... I don't understand what adding the suggested line does. I am ok adding it - but I am either understanding it wrong, or it is not needed :)
-> This would mark the ubuntu-latest image to have rust 1.70.0 installed? Probably not really installing that (as for that the dockerfile would need to be changed), but marking it for the reader of the ci-file?

MSRV=1.60 means exactly what you said.
But cargo-msrv does both: Using bisect to find the minimal rust version, by installing multiple rust versions -> this I did locally only.  
Or verifying that a set version is (still) correct, by installing only that rust version and executing cargo check. -> (only) this is done in ci.

The line 'rust-version' in frb_codegen/cargo.toml is hindering anyone to cargo install frb_codegen with a lower rust version. Cargo-msrv does nothing here.

The test in the ci with cargo msrv only checks if changes in frb_codegen need a higher rust version.

For example: If a change would require 1.75.0, instead of the version 1.74.0 set in cargo.toml, the ci run would fail.
But a user with rust 1.74.0 could still run cargo install frb_codegen, which would fail at compiling, but not tell that the MSRV is higher.  
Only after manually updating the cargo.toml to rust-version-1.75.0 the ci would be green again, and that user would get the error that his rust version is too old for cargo install frb_codegen.

Sorry for being so verbose - I am sure you got almost all already :)
So, what is your suggested line in ci.yaml doing?

Reply:
> -> This would mark the ubuntu-latest image to have rust 1.70.0 installed? Probably not really installing that (as for that the dockerfile would need to be changed), but marking it for the reader of the ci-file?

It means, IMHO, add a new test (so called "matrix") with condition rust 1.70 (while existing tests have stable, nightly, etc)
For more info, maybe refer to github actions' documentations. 

> Or verifying that a set version is (still) correct, by installing only that rust version and executing cargo check. -> (only) this is done in ci.

This looks interesting! However, if I understand correctly, using this part of cargo-msrv seems to have the same effect as adding the two lines suggested above. If so, we can choose the simpler method to avoid future maintainence burden :) But cargo-msrv for local testing is quite good.

> The line 'rust-version' in frb_codegen/cargo.toml is hindering anyone to cargo install frb_codegen with a lower rust version. Cargo-msrv does nothing here.

This LGTM!

Reply:
> It means, IMHO, add a new test (so called "matrix") with condition rust 1.70 (while existing tests have stable, nightly, etc) For more info, maybe refer to github actions' documentations.
> 
Ah, yes, I see that now :)
> > Or verifying that a set version is (still) correct, by installing only that rust version and executing cargo check. -> (only) this is done in ci.
> 
> This looks interesting! However, if I understand correctly, using this part of cargo-msrv seems to have the same effect as adding the two lines suggested above. If so, we can choose the simpler method to avoid future maintainence burden :) But cargo-msrv for local testing is quite good.

Yes - if 
```
      # execute
      - run: ./frb_internal test-rust --coverage`
```
is ONLY executing tests in frb_codegen (not easy to figure out). If it runs other tests as well we would run them on that version as well.

I think your suggestion works - but maintenance would get worse (?): I you mean by "more maintenance" that we depend on cargo-msrv, then yes, your alternative would remove that dependency.

But, if I am not wrong, that more than just frb_codegen tests are run here, then:  
One could install frb_codegen with binstall - and if the generated crates have a lower MSRV, all would be ok for that user. But the ci would fail here.  

Thus, we fail in a situation where everything would be ok, and the ci takes more time by running this scenario as well. And maybe a user needs to use an older rust toolchain.

However, `test-rust` is limited to frb_codegen only, I agree with your suggestion!
If not - well, my scenario is an unlikely edge-case. So, if my assumptions are correct, but you would still prefer independence from cargo-msrv over the drawbacks mentioned, I will implement your suggestion anyways :)

Reply:
> is ONLY executing tests in frb_codegen (not easy to figure out). If it runs other tests as well we would run them on that version as well.

Runs test for all rust packages inside this git repository indeed. But it looks good if we ensure everything has the same MSRV, which simplifies things.

In addition, not only frb_codegen, but also the MSRV for frb_rust (the runtime library) is quite important. We surely do not want a low frb_codegen MSRV, while super high frb_rust MSRV!

Reply:
> Runs test for all rust packages inside this git repository indeed. But it looks good if we ensure everything has the same MSRV, which simplifies things.
True - I can look into adding the rust-version to all tomls, where they are generated.
> 
> In addition, not only frb_codegen, but also the MSRV for frb_rust (the runtime library) is quite important. We surely do not want a low frb_codegen MSRV, while super high frb_rust MSRV!
But the other way around might be desired: frb_codegen needs 1.74.0. Let's assume the rest is fine with 1.60.0. A user can get around with not having 1.74.0 by `cargo binstall frb_codegen`. The rest needs to be compiled, so anything below 1.60.0 would not work, but 1.60.0 would be fine. Maybe he has a reason for not upgrading to 1.74.0 (any of his dependencies not working with that version). I just want to illustrate this point so the tradeoff is clear.

So, how I see it:
- Variant A: cargo msrv checks the MSRV of frb_codegen only
- Variant B: all rust packages in this repo are run against a specific rust version

Contra-points:
- Variant A: 
  - ci depends on cargo-msrvg
- Variant B:
  - all rust packages of this repo need have the same MSRV, which is the highest among them. (Actually: All compiled packages vs binstallable (which is frb_codegen only)
    - users are potentially forced to use a higher rust version, if the coiled packages can have a lower MSRV than the binstallable. 
  - ci takes longer, as all packages are tested in the new image
  - maintenance is higher, as the MSRV needs to be updated in two places: cargo.toml and the ci image definition

Do you see anything else or disagree with that? I think variant A is better - but I will implement whatever you think is best. After all, both variant's contra points are not very strong, so it doesn't matter much - and it is your project :)

Reply:
> ci takes longer, as all packages are tested in the new image

If we are worried, maybe we can add a flag to the script to enable only one package vs all packages.
But generally speaking, from my observations, CI for *this* part is quite fast, at least much faster than ios tests etc.

So are we really having very different MSRVs here? Our code is quite standard, so I guess indeed we will have a pretty good MSRV such that there is no worry at all about "frb_codegen different MSRV than frb_rust".

Reply:
> If we are worried, maybe we can add a flag to the script to enable only one package vs all packages. But generally speaking, from my observations, CI for _this_ part is quite fast, at least much faster than ios tests etc.

Sounds good!

> So are we really having very different MSRVs here? Our code is quite standard, so I guess indeed we will have a pretty good MSRV :) such that there is no worry at all about "frb_codegen different MSRV than frb_rust".

I think at the moment we have that: frb_codegen needs 1.74.0 - that was the initial issue.
The generated code and runtime is fine with a lower MSRV (I checked for `dart_minimal`, which is fine with 1.72.1, and frb_rust, which is fine with  1.70.0.).

I am most concerned about the additional maintenance needed in the ci file (though that is minimal admittedly):  
If  the ci fails because a higher MSRV is needed, only the cargo.toml has to be updated.  
In Variant B the toml and the ci.yaml need both an update. Not a big deal - but some additional thing to keep in mind.

Reply:
> Not a big deal - but some additional thing to keep in mind.

Yes I agree it is not a huge problem :)

Reply:
Sounds like you prefer Variant B :) Good, will change this PR later.

I will set the MSRV in frb_rust as well.  
Setting it in the examples requires a bit more effort to make that properly - as this should be done through code generation.  
But I think this is not needed - as a user needs to have the MSRV for frb_rust anyways, and I assume there is no way that the created/integrated project would have something that requires a higher MSRV.
So, I will skip that, ok?

Reply:
No need for examples! Since they are just examples :)

Reply:
It turns out that the changes are not that trivial ... 

1. setting the MSRV in ci.yaml as a variable is challenging - not sure if it does the trick in the end (see the diff).
2. the CI fails, as there are two identical jobs now (`Test :: Rust :: ubuntu-latest`). Logs like the version provided is not picked up.
3. The ci should fail, as I intentionally set the MSRV under the one specified - but it doesn't. Probably the version is indeed not picked up correctly.

I don't really want to hardcode the version without a variable, so one understands this is for MSRV. Will do so now, just to see if that is the only issue. But if this takes some more digging into to fix I prefer to save some time and revert back to Variant A.

Reply:
It works now :) Please look into the code. The MSRV is hardcoded, but I commented it so that it should be clear what to do if it fails in the future.
I left the implementation for Variant A inside as well - but commented.
I know, the code is in git as well - but I bet that just in a few weeks it would be hard to digg for that.

I ran some test ci runs with intentionally wrong version numbers - all works as expected.

Please let me know if I can do anything else regarding this PR :)